### PR TITLE
run_tests.sh: add --clear-cache flag to clear all test-related cache

### LINKED ISSUE #1017: Container images are not removed in `setup.sh --completely-remove`
**Description:**

After completely removing kw, images from integration tests are not removed.

**How to Reproduce:**

1. With unstable kw installed, execute in kw source tree
```bash
./run_tests --integration
```
2. After it finishes, you can see that the images are installed runnning
```bash
podman image ls
```
3. Then, completely remove kw with
```bash
./setup.sh --completely-remove
```
4. Run this command again to confirm that the images are still there:
 ```bash
podman image ls
```
You should see something like
```
$ podman image ls                                                                                                                  [19:43:36]
REPOSITORY                   TAG         IMAGE ID      CREATED       SIZE
localhost/kw-fedora          latest      eef06db6345a  6 hours ago   1.16 GB
localhost/kw-debian          latest      40227dd4c41a  6 hours ago   2.39 GB
localhost/kw-archlinux       latest      e912e0d8d8dd  6 hours ago   1.24 GB
docker.io/library/debian     latest      a6916e41aa87  20 hours ago  121 MB
docker.io/library/archlinux  latest      69f38d8f6347  10 days ago   446 MB
docker.io/library/fedora     latest      8404925a71fd  6 weeks ago   183 MB
```

**Expected behavior:**

It is expected that the images are gone after completely removing kw from system.

**System information:**

 - OS: Linux fedora 6.6.9
 - kw Version: 
``` 
alpha
Branch: unstable
Commit: 1764e42
```
 - Bash Version 5.2.21

### PR DESCRIPTION
**Pull request by uwla**
Currently, this new option clears cached container images.

Fixes #1017

### CODE REVIEW

File: run_tests.sh
Comment (rodrigosiqueira) [inline run_tests.sh:187]: Use single quotes; also, add this option in the help output.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

File: run_tests.sh
Comment (davidbtadokoro) [inline run_tests.sh:186]: By the way the conditionals are, the following cases are acceptable
```bash
./run_tests.sh --unit --clear-cache
./run_tests.sh --integration --clear-cache
./run_tests.sh --clear-cache
```
Which one are you aiming to?

IMHO, I think it makes sense to tie the option to `--integration` like in the the second case and maybe have
```bash
./run_tests.sh --integration clear-cache
```
just to keep the pattern of set by the `test`, `list`, and `help` options that come after `--unit`/`--integration`.

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hey @uwla, and thank you so much this change.

Tested it and it seems to do what it proposes to. In any case, I came upon the errors from podman shown below when removing some images. I think that this may be a problem on my end, but can you double check the behavior, just to be sure?

Screenshot from 2024-01-17 15-19-35

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:195]: I think we should make this docker prefix path a global read-only variable because, in the future, if we want to use something different, we can just change the global variable.

**Review by rodrigosiqueira — COMMENTED** [review]
I can confirm that I had a similar issue as the one reported by @davidbtadokoro .

Screenshot from 2024-01-18 17-45-15

File: tests/integration/utils.sh
Comment (uwla) [reply tests/integration/utils.sh:195]: True, indeed

File: run_tests.sh
Comment (uwla) [reply run_tests.sh:187]: ok

File: run_tests.sh
Comment (uwla) [reply run_tests.sh:124]: Ok

File: run_tests.sh
Comment (uwla) [reply run_tests.sh:186]: That is a good idea: to have cached cleared on a scope basis.

**Comment by uwla** [comment]
> Hey @uwla, and thank you so much this change.
> 
> Tested it and it seems to do what it proposes to. In any case, I came upon the errors from podman shown below when removing some images. I think that this may be a problem on my end, but can you double check the behavior, just to be sure?
> 
> Screenshot from 2024-01-17 15-19-35

As the podman's output points out, the kw images are being used by other images. I had experienced this behavior and deleted intermediary images before. The problem is this: it seems Podman generates duplicated entries of the built images; you can verify that by `podman image ls -a` and see the images seem to be equal.

One way we can address that is to force remove the images.

**Comment by davidbtadokoro** [comment]
> One way we can address that is to force remove the images.

This may be a way to do this. Adding the flag `--force` to the command
```bash
podman image rm --force "$img"
```
seems to do the trick.

However, we need to be sure that doing this won't mess any other podman image besides the ones related to kw's integration tests :grimacing:

**Comment by uwla** [comment]
> However, we need to be **really** sure that doing this won't mess any other podman image besides the ones related to kw's integration tests 😬

I have two suggestions:

1. Interactvely ask the user if KW should proceed and force the removal of the images. This is the easiest way, but I suspected the other "duplicated" podman images would remain (this is problem of duplicating entries of image layers is related to podman, not KW)
2. Figure out a way to inspect which other images are dependent on the one we are trying to remove, and if they are just duplicated entries, we remove them first.

**Comment by uwla** [comment]
I cleared all podman images and container in my system, then ran the integration tests, which built the images locally. Here is the output:

```
[kworkflow][ unstable][$⇡] ❯ podman image ls                             
REPOSITORY                   TAG         IMAGE ID      CREATED        SIZE
localhost/kw-fedora          latest      dba3438ea2ed  2 minutes ago  1.17 GB
localhost/kw-debian          latest      668dde0f82f7  5 minutes ago  2.39 GB
localhost/kw-archlinux       latest      beab320280c0  7 minutes ago  1.34 GB
docker.io/library/archlinux  latest      69f38d8f6347  3 weeks ago    446 MB
docker.io/library/debian     latest      2a033a8c6371  5 weeks ago    121 MB
docker.io/library/fedora     latest      8404925a71fd  8 weeks ago    183 MB
[kworkflow][ unstable][$⇡] ❯ podman image ls -a                        
REPOSITORY                   TAG         IMAGE ID      CREATED        SIZE
localhost/kw-fedora          latest      dba3438ea2ed  2 minutes ago  1.17 GB
<none>                       <none>      e76b911a33d6  3 minutes ago  509 MB
<none>                       <none>      b916531e6fee  3 minutes ago  509 MB
localhost/kw-debian          latest      668dde0f82f7  5 minutes ago  2.39 GB
<none>                       <none>      6353772123c4  7 minutes ago  267 MB
<none>                       <none>      c9c5cc748609  7 minutes ago  267 MB
localhost/kw-archlinux       latest      beab320280c0  8 minutes ago  1.34 GB
<none>                       <none>      bc960f2cb82f  9 minutes ago  755 MB
<none>                       <none>      83b04e8e5557  9 minutes ago  755 MB
docker.io/library/archlinux  latest      69f38d8f6347  3 weeks ago    446 MB
docker.io/library/debian     latest      2a033a8c6371  5 weeks ago    121 MB
docker.io/library/fedora     latest      8404925a71fd  8 weeks ago    183 MB
```

The images marked by `<none>` are byproducts of the building process and can be removed without any loss. Really, they can be removed and the other images will work fine. So, we could do:

```
$ podman image ls -a | grep '<none>' | awk '{ print $3 }' | xargs podman image rm -f
```

However, I cannot ensure removing images marked by `<none>` won't remove images which the user actually wants to keep.

**Comment by uwla** [comment]
I also noticed all these `temporary` images have the `-tmp` prefix in their names:

```
❯ podman image ls -aq | xargs podman inspect -f "{{.NamesHistory}}"
[localhost/kw-fedora:latest]
[docker.io/library/c1b1a7ee84e9430ab72ca01da8d8589c21cf2495a6144e5218dda596990c8b73-tmp:latest]
[docker.io/library/51bda72fc5b773d2aec762bf4cea2b1532c4fd6ee59575bed7c202230d30cce1-tmp:latest]
[localhost/kw-debian:latest]
[docker.io/library/0f4612409cf1443e4b462185c3942187a3753692ff6263bfa68b2acef9d59cb3-tmp:latest]
[docker.io/library/5b22c6a784bbbe871b20f772cc184815c5e5f73506bf09f53d55dfd783fd0400-tmp:latest]
[localhost/kw-archlinux:latest]
[docker.io/library/f378a37dd967a15f6bc8b67bf8a0ab117c6e4875581cfbc3fa6b6903d13790c9-tmp:latest]
[docker.io/library/7bfa6850b870c614a20211ba34ed081338e53d3e5d92f47caa8d4565d18eaa96-tmp:latest]
[docker.io/library/archlinux:latest]
[docker.io/library/debian:latest]
[docker.io/library/fedora:latest]
```

We could assume if a image is marked by `-tmp`, it is a temporary image, byproduct of the building process, and can be discard without any problems. I believe this is the way to go.

If I find a better solution, I'll post it here.

**Comment by uwla** [comment]
Ok, I found out a better solution but this will require some extra string parsing.

We can see the podman images, including the temporary images, share layers:

```
 podman image ls -a | grep -v "IMAGE ID" | awk '{ print $3}' | xargs podman inspect -f "{{.RootFS.Layers}}" | sort
[sha256:53f86715cdbaf45ddf227b5582e881e3ad0d5eb056eda3b0094a66f7cf0c0f1f]
[sha256:53f86715cdbaf45ddf227b5582e881e3ad0d5eb056eda3b0094a66f7cf0c0f1f sha256:a0f125ca6de8ea112b7a6fe7c99afe5b0cac482a666b8b1a0b637d18902358b1]
[sha256:53f86715cdbaf45ddf227b5582e881e3ad0d5eb056eda3b0094a66f7cf0c0f1f sha256:a0f125ca6de8ea112b7a6fe7c99afe5b0cac482a666b8b1a0b637d18902358b1 sha256:3155923fa362a69998dd8afadf32efe9571ede98920a38c2f659bee46b0f7fec]
[sha256:53f86715cdbaf45ddf227b5582e881e3ad0d5eb056eda3b0094a66f7cf0c0f1f sha256:a0f125ca6de8ea112b7a6fe7c99afe5b0cac482a666b8b1a0b637d18902358b1 sha256:3155923fa362a69998dd8afadf32efe9571ede98920a38c2f659bee46b0f7fec sha256:72c62a91e48f8188ff0acb936ae9ebabcf1477e26ea4220860936016ae591a1d]
[sha256:6de0fb71036567b0339788eaa924723efbd5be22f9683b8aeb3b3c23d4e41a9d sha256:d822bba4f032dfd2df5b57457840d3ef708ec0f592b3bc29cc2cba6cf271de23]
[sha256:6de0fb71036567b0339788eaa924723efbd5be22f9683b8aeb3b3c23d4e41a9d sha256:d822bba4f032dfd2df5b57457840d3ef708ec0f592b3bc29cc2cba6cf271de23 sha256:a0030b74034672418b6f47e34735be3c31e1591a87b601afaf695a79e279610b]
[sha256:6de0fb71036567b0339788eaa924723efbd5be22f9683b8aeb3b3c23d4e41a9d sha256:d822bba4f032dfd2df5b57457840d3ef708ec0f592b3bc29cc2cba6cf271de23 sha256:a0030b74034672418b6f47e34735be3c31e1591a87b601afaf695a79e279610b sha256:3155923fa362a69998dd8afadf32efe9571ede98920a38c2f659bee46b0f7fec]
[sha256:6de0fb71036567b0339788eaa924723efbd5be22f9683b8aeb3b3c23d4e41a9d sha256:d822bba4f032dfd2df5b57457840d3ef708ec0f592b3bc29cc2cba6cf271de23 sha256:a0030b74034672418b6f47e34735be3c31e1591a87b601afaf695a79e279610b sha256:3155923fa362a69998dd8afadf32efe9571ede98920a38c2f659bee46b0f7fec sha256:02a0ac4c6157eb3f57f23838293c1668f4205fbbe2be38716f99313885f708de]
[sha256:ae134c61b154341a1dd932bd88cb44e805837508284e5d60ead8e94519eb339f]
[sha256:ae134c61b154341a1dd932bd88cb44e805837508284e5d60ead8e94519eb339f sha256:c2ff07d5b87862d11b7e954cdb1a261fcc3b8f53f83b9b6b11a8ce74988c0f04]
[sha256:ae134c61b154341a1dd932bd88cb44e805837508284e5d60ead8e94519eb339f sha256:c2ff07d5b87862d11b7e954cdb1a261fcc3b8f53f83b9b6b11a8ce74988c0f04 sha256:3155923fa362a69998dd8afadf32efe9571ede98920a38c2f659bee46b0f7fec]
[sha256:ae134c61b154341a1dd932bd88cb44e805837508284e5d60ead8e94519eb339f sha256:c2ff07d5b87862d11b7e954cdb1a261fcc3b8f53f83b9b6b11a8ce74988c0f04 sha256:3155923fa362a69998dd8afadf32efe9571ede98920a38c2f659bee46b0f7fec sha256:91dd6e2a58e3d7ed997362d41b005ff03080eca25da14226252595e4b1a3c1e3]
```

We can then delete all podman images whose layers are a prefix of the KW images.

```
tmp image:  abc def ghi
kw archlinux image: abc def ghi jkl mno
other image: abc def ghi poq
```

In the example above, `tmp` is a prefix of `kw archlinux image`, but `other image` is not because `poq` layer is not present in `kw archlinux image`.

Using this heuristic, we can be confident only the byproduct images will be removed along with kw images.

Please, give me our opinions @davidbtadokoro or @rodrigosiqueira.

**Comment by uwla** [comment]
Ok, I'll implement the solution soon and update the PR

**Comment by uwla** [comment]
@davidbtadokoro please review it again

File: tests/integration/utils.sh
Comment (davidbtadokoro) [inline tests/integration/utils.sh:211]: A cleaner way to do this would be
```bash
var=$(printf '%s' "$img_layers" | grep --perl-regexp "^${tmp_img_layers}")
if [[ -n "$var" ]]; then
 ...
```
Of course, change the name `var` to a more adequate one.

I say that this is cleaner, because we don't need to redirect the output to `/dev/null`, reading it from left to right (beginning to end) makes more sense, and, most importantly, it follows our coding style.

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hey @uwla, and thanks for the update fixing the issue me and @rodrigosiqueira reported!

Talking about the changes requested, one relates to the commit message, and the others are mainly about coding style and improving readability.

In terms of solving the reported issue, your update really did the trick, as it deletes the temporary dangling images created by the integration tests, and only those! However, I got some strange errors that don't seem to affect the outcome. See the screenshots below:

Screenshot from 2024-01-31 00-00-07
Screenshot from 2024-01-31 00-01-04
Screenshot from 2024-01-31 00-01-15

**Comment by uwla** [comment]
@davidbtadokoro I made the requested changes

File: run_tests.sh
Comment (davidbtadokoro) [inline run_tests.sh:197]: Use single-quotes instead of double-quotes when there is no expansion.

**Review by davidbtadokoro — COMMENTED** [review]
Hi @uwla, and many many thanks for the update! Everything looks legit to me, from both the coding style and behavior standpoints :+1:. It also fixed those error messages about failing to inspect. There is a small nitpicky comment about single quotes that I left just for documentation, but I can fix this myself, no problems. Nevertheless, great contribution and great effort all around :smiley: 

@rodrigosiqueira, as you also reviewed this, I'll wait for your green light to merge this PR.

**Comment by rodrigosiqueira** [comment]
Hi @uwla , it looks like we have a conflict in this PR. Could you update it against the latest code from the unstable branch?

**Comment by uwla** [comment]
yes, I'll

**Comment by davidbtadokoro** [comment]
Hi @uwla, any update here?

**Comment by davidbtadokoro** [comment]
Thanks for resolving the conflicts @uwla. @rodrigosiqueira, I'll wait for your greenlight to merge it.

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:250]: In my case, I had a failure that looked like this:

```
Removing podman docker.io/library/debian image.
Error: image used by c23446c400585541e620124d8202d5ed41dac46b5a7b0ee9b92907ae8d2406f3: image is in use by a container: consider listing external containers and force-removing image
(253): Failed to remove the images.
```

In the output, it suggests listing external containers and force-removing. Could you add these steps as part of the error message? I mean, add the podman command to list external containers and the command that forces the removal.

Comment (rodrigosiqueira) [inline tests/integration/utils.sh:250]: Just in case, check if `$@` has at least one parameter.

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:260]: Just in case, check if `$@` has at least one parameter.

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:270]: Just in case, check if `$@` has at least one parameter.

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:263]: To make debugging easier in case of failure, add the command you tried to execute as part of the error message. Something like: `kw failed to execute the command: podman image ls SOMETHING SOMETHING`

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:273]: To make debugging easier in case of failure, add the command you tried to execute as part of the error message. Something like: `kw failed to execute the command: podman inspect ls SOMETHING SOMETHING`

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:247]: Add function documentation on top, in particular, explaining that "$@" is a set of flags that will be passed to Podman. Do the same for all the other functions.

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:225]: I'm wondering if we can fail in this part... maybe we can run this command before this while, save the output in a variable/array, and use it in this while loop if everything is fine. This can make the code more reliable and easier to debug in case something happens.

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:187]: This variable looks very generic, and it is hard to guess what it does from its name. Fwiu, you want to get part of the layer image name, right? If so, rename this variable to something that expresses that idea.

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:185]: This name looks a little bit generic; maybe use something like podman_image_list or something like that.

File: tests/integration/utils.sh
Comment (rodrigosiqueira) [inline tests/integration/utils.sh:186]: Same here... this variable name is a little bit generic.

File: run_tests.sh
Comment (rodrigosiqueira) [inline run_tests.sh:135]: If there is nothing to clean, just say that; otherwise, you can give the wrong idea to the user. Maybe say, `Skip unit test clean` or `Clear unit test is not necessary` or anything like that.

File: run_tests.sh
Comment (rodrigosiqueira) [inline run_tests.sh:21]: Don't forget to update documentation/content/tests.rst with this new option.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @uwla , thanks a lot for the update, and sorry for my delay in reviewing it. You can find some of my comments in this PR.

File: run_tests.sh
Comment (uwla) [reply run_tests.sh:21]: ok

File: tests/integration/utils.sh
Comment (uwla) [reply tests/integration/utils.sh:263]: ok

**Comment by uwla** [comment]
Everything should be working fine by now.

@rodrigosiqueira and @davidbtadokoro,  I tested it and worked on my side, but I also ask you to test it locally, because we do not have automated integration tests yet.

**Comment by davidbtadokoro** [comment]
Nice @uwla ! I'll take a look at it : )

File: run_tests.sh
Comment (davidbtadokoro) [reply run_tests.sh:197]: Make it single-quotes here.

File: documentation/content/tests.rst
Comment (davidbtadokoro) [inline documentation/content/tests.rst:63]: I think your formatter changed these two paragraphs. Revert them to the previous state.

File: tests/integration/utils.sh
Comment (davidbtadokoro) [inline tests/integration/utils.sh:148]: There is no `flags` argument, so remove it from the function doc.

File: tests/integration/utils.sh
Comment (davidbtadokoro) [inline tests/integration/utils.sh:191]: Fix `Gemove` typo.

File: tests/integration/utils.sh
Comment (davidbtadokoro) [inline tests/integration/utils.sh:194]: Fix `iamge` typo.

File: tests/integration/utils.sh
Comment (davidbtadokoro) [inline tests/integration/utils.sh:301]: FWIU, this `flags` argument are podman options that will be passed through `$@` (along with the `images`) to the podman command. Instead of separating the documentation of these two arguments, I think it is better to describe them as one (all arguments passed to the function), and explicitly mentioning that it is composed of (first) podman options and (second) image names/IDs.

File: tests/integration/utils.sh
Comment (davidbtadokoro) [inline tests/integration/utils.sh:305]: Use double-quotes around `$#`.

File: tests/integration/utils.sh
Comment (davidbtadokoro) [inline tests/integration/utils.sh:320]: Same comment as the one I made on the `image_ls` function.

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hi @uwla, and thanks for the update! We are really close to the perfect point!

Left just some simple comments this time. Also, a comment that didn't fit the inline review: remove the final dot in the commit message subject.

**Comment by uwla** [comment]
@davidbtadokoro I've made the requested changes. Hopefully it passes all checks now

**Comment by rodrigosiqueira** [comment]
Hi @uwla , thanks a lot for your update. I decided to merge this change into the unstable branch.
Thanks a lot.

**Pull request closed without merging**
